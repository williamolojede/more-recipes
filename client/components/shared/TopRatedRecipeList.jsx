import React from 'react';

import { topRatedRecipeListPropTypes } from '../../config/proptypes';

import RecipeCard from './RecipeCard/RecipeCard.jsx';

const TopRatedRecipeList = props => (
  <ul className="top-rated__list row">
    {props.recipes.map(recipe => (
      <li className="top-rated__item col s12 m4 l4 x3" key={recipe.id}>
        <RecipeCard recipe={recipe} />
      </li>
    ))}
  </ul>
);

TopRatedRecipeList.propTypes = topRatedRecipeListPropTypes;

export default TopRatedRecipeList;
