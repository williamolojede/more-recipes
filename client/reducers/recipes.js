/**
 * Reducer for All Recipe
 */

import { RECIEVE_TOP_RATED_RECIPE } from '../actions/types';

const recipes = (state = [], action) => {
  switch (action.type) {
    case RECIEVE_TOP_RATED_RECIPE:
      return action.recipes;
    default:
      return state;
  }
};


export default recipes;
