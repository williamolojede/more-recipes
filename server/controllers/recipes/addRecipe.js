import { Recipe } from '../../models/index';

const addRecipe = (req, res, next) => {
  const { name, description, img_url, ingredients, instructions } = req.body;
  Recipe.create({
    name,
    description,
    img_url,
    ingredients,
    instructions,
    owner: req.userID
  })
    .then(recipe => res.status(201).json({ recipe, message: 'success' }))
    .catch((error) => {
      const err = new Error(error);
      err.status = 500;
      next(err);
    });
};

export default addRecipe;
