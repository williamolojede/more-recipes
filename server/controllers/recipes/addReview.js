import { Review } from '../../models/index';
import helpers from '../../helpers';
import { findReviews } from './getReviews';

const addReview = (req, res, next) => {
  const { content } = req.body;
  const { userID: userId } = req;
  const { id: recipeId } = req.params;

  Review.create({ userId, recipeId, content })
    .then(() => findReviews(req))
    .then(({ rows: reviews, pagination }) => res.status(201).send({
      status: 'success',
      reviews,
      pagination,
      message: 'Your review has been recorded'
    }))
    .catch(error => helpers.systemErrorHandler({
      msg: error.errors[0].message,
      code: 400
    }, next));
};

export default addReview;

