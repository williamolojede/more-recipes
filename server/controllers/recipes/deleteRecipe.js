import { Recipe } from '../../models/index';

const deleteRecipe = (req, res, next) => {
  const { userID } = req;
  // const decoded = jwt.decode(token);
  Recipe.findById(req.params.id)
    .then((recipe) => {
      if (userID !== recipe.owner) {
        const err = new Error('Not authorized to delete this recipe');
        err.status = 403;
        return next(err);
      }
      recipe.destroy()
        .then(() => res.status(200).send({ message: 'success' }))
        .catch(err => console.error(err));
    })
    .catch((error) => {
      const err = new Error(error);
      err.status = 401;
      next(err);
    });
};

export default deleteRecipe;
