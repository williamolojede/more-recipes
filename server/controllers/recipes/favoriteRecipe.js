import { Favorite } from '../../models/index';

const favoriteRecipe = (req, res, next) => {
  const { userID, recipeOwner } = req;
  const recipeId = req.params.id;

  // ensures owner doesn't favorite their own recipe
  if (recipeOwner === userID) {
    const err = new Error('you are not allowed to vote on your own recipe');
    err.status = 403;
    return next(err);
  }

  Favorite.findAll({ where: { recipeId } })
    .then((favorites) => {
      // check if the favorite table is empty i:e its the first favorite for the recipe
      if (favorites.length === 0) {
        return Favorite.create({
          userId: userID,
          recipeId
        })
          .then(() => res.status(200).send({ status: 'success', message: 'recipe added to your favorite list' }))
          .catch((error) => {
            const err = new Error(error);
            err.status = 500;
            return next(err);
          });
      }
      // get list of all userId that has already favorited the recipe
      const alreadyfavorited = [];
      favorites.forEach((el) => {
        alreadyfavorited.push(el.dataValues.userId);
      });
      const userFavorite = favorites.filter(el => el.dataValues.userId === userID)[0];
      // if current user already favorited => remove
      if (alreadyfavorited.includes(userID)) {
        return Favorite.findById(userFavorite.dataValues.id)
          .then(favorite => favorite.destroy())
          .then(() => res.status(200).send({ message: 'recipe removed from your favorite list', status: 'success' }));
      }
      Favorite.create({
        userId: userID,
        recipeId
      })
        .then(() => res.status(200).send({ status: 'success', message: 'recipe added to your favorite list' }))
        .catch((error) => {
          const err = new Error(error);
          err.status = 500;
          return next(err);
        });
    });
};

export default favoriteRecipe;
