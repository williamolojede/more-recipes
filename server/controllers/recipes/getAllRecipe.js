import { Recipe, Review, User } from '../../models/index';
import systemErrorHandler from '../../helpers/systemErrorHandler';
import Paginate from '../../helpers/paginate';

const getAllRecipe = (req, res, next) => {
  const { sort, order } = req.query;

  //  if query is passed
  if (sort && order) {
    // if the wrong sort order is passed
    if (order !== 'ascending' && order !== 'descending') {
      const err = new Error('invalid sort order');
      err.status = 400;
      return next(err);
    }
    return Recipe.findAll({
      order: [['upVoteCount', 'DESC']],
      include: [
        { model: Review, as: 'reviews' },
        { model: User, attributes: ['id', 'username', 'fullname'] }
      ]
    })
      .then((topRecipes) => {
        const {
          recipes,
          metaData
        } = (new Paginate(topRecipes, parseInt(req.query.limit, 10)))
          .getRecipesForPage(parseInt(req.query.page, 10));
        res.status(200).send({ recipes, metaData, message: 'success' });
      })
      .catch(error => systemErrorHandler(error, next));
  }
  Recipe.findAll({
    include: [
      { model: Review, as: 'reviews' },
      { model: User, attributes: ['id', 'username', 'fullname'] }
    ]
  })
    .then(recipes => res.status(200).send({ recipes, message: 'success' }))
    .catch(error => systemErrorHandler(error, next));
};
export default getAllRecipe;
