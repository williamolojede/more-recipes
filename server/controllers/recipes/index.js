import createRecipe from './addRecipe';
import getAllRecipe from './getAllRecipe';
import updateRecipe from './updateRecipe';
import deleteRecipe from './deleteRecipe';

const recipesController = {
  getAllRecipe,
  createRecipe,
  updateRecipe,
  deleteRecipe
};


export default recipesController;
