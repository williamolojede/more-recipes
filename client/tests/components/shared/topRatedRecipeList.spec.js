import React from 'react';
import TopRatedRecipeList from '../../../components/shared/TopRatedRecipeList';
import recipes from '../../../__mocks__/recipes.mock';

function setup() {
  const props = {
    recipes,
  };
  const wrapper = shallow(<TopRatedRecipeList {...props} />);

  return {
    wrapper
  };
}

describe('Top Rated Recipes List', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('RecipeCard').length).toBe(recipes.length);
  });
});
