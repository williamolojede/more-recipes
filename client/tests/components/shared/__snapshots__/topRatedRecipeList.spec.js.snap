// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Rated Recipes List should render with right amount of elements 1`] = `
<ul
  className="top-rated__list row"
>
  <li
    className="top-rated__item col s12 m6 l4 x4"
    key="3"
  >
    <RecipeCard
      recipe={
        Object {
          "User": Object {
            "fullname": "John Bull",
            "id": 8,
            "username": null,
          },
          "createdAt": "2017-11-03T09:51:16.180Z",
          "description": "This is one of my 'go to' recipes and it takes only about 15 minutes from start to finish. The only prep work really is mashing the garlic. The sauce is great on any pasta, but I prefer angel hair or spaghetti.",
          "downVoteCount": 0,
          "favoriteCount": 0,
          "id": 3,
          "img_url": "https://firebasestorage.googleapis.com/v0/b/more-recipes.appspot.com/o/recipes%2F1509702436404?alt=media&token=da3a7ac8-58d3-46da-a71d-c2c1c0cbdfeb",
          "ingredients": Array [
            "2 tablespoons olive oil",
            "3 cloves garlic",
            "2 tablespoons capers",
            "1 tablespoon anchovy paste",
            "1 teaspoon red pepper flakes",
            "1 (24 ounce) jar tomato pasta sauce",
            "1/4 cup water",
            "1 cup good-quality crisp white wine",
            "2 (6.5 ounce) cans chopped clams, drained with juice reserved",
            "salt and pepper to taste",
            "1 pound dry spaghetti",
            "1/4 cup freshly grated Parmesan cheese",
            "Chopped fresh basil leaves or parsley",
          ],
          "instructions": Array [
            "Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato",
            "Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.",
            "Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.",
          ],
          "name": "Chef John's Spaghetti with Red Clam Sauce",
          "owner": 8,
          "reviews": Array [
            Object {
              "User": Object {
                "fullname": "John Bull",
                "id": 8,
                "username": null,
              },
              "content": "gwbvgfda",
              "createdAt": "2017-11-04T07:28:31.732Z",
              "id": 5,
              "recipeId": 3,
              "updatedAt": "2017-11-04T07:28:31.732Z",
              "userId": 8,
            },
          ],
          "upVoteCount": 1,
          "updatedAt": "2017-11-09T11:36:49.091Z",
          "viewCount": 1,
        }
      }
    />
  </li>
  <li
    className="top-rated__item col s12 m6 l4 x4"
    key="106"
  >
    <RecipeCard
      recipe={
        Object {
          "User": Object {
            "fullname": "Email Mail",
            "id": 11,
            "username": null,
          },
          "createdAt": "2017-11-11T09:07:36.283Z",
          "description": "This is one of my 'go to' recipes ",
          "downVoteCount": 0,
          "favoriteCount": 0,
          "id": 106,
          "img_url": "http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg",
          "ingredients": Array [
            "2 tablespoons olive oil",
            "3 cloves garlic",
            "2 tablespoons capers",
            "1 tablespoon anchovy paste",
            "1 teaspoon red pepper flakes",
            "1 (24 ounce) jar tomato pasta sauce",
            "1/4 cup water",
            "1 cup good-quality crisp white wine",
            "2 (6.5 ounce) cans chopped clams, drained with juice reserved",
            "salt and pepper to taste",
            "1 pound dry spaghetti",
            "1/4 cup freshly grated Parmesan cheese",
            "Chopped fresh basil leaves or parsley",
          ],
          "instructions": Array [
            "Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato",
            "Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.",
            "Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.",
          ],
          "name": "Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref",
          "owner": 11,
          "reviews": Array [],
          "upVoteCount": 0,
          "updatedAt": "2017-11-11T09:07:36.283Z",
          "viewCount": 0,
        }
      }
    />
  </li>
  <li
    className="top-rated__item col s12 m6 l4 x4"
    key="101"
  >
    <RecipeCard
      recipe={
        Object {
          "User": Object {
            "fullname": "Email Mail",
            "id": 11,
            "username": null,
          },
          "createdAt": "2017-11-11T09:07:34.864Z",
          "description": "This is one of my 'go to' recipes ",
          "downVoteCount": 0,
          "favoriteCount": 0,
          "id": 101,
          "img_url": "http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg",
          "ingredients": Array [
            "2 tablespoons olive oil",
            "3 cloves garlic",
            "2 tablespoons capers",
            "1 tablespoon anchovy paste",
            "1 teaspoon red pepper flakes",
            "1 (24 ounce) jar tomato pasta sauce",
            "1/4 cup water",
            "1 cup good-quality crisp white wine",
            "2 (6.5 ounce) cans chopped clams, drained with juice reserved",
            "salt and pepper to taste",
            "1 pound dry spaghetti",
            "1/4 cup freshly grated Parmesan cheese",
            "Chopped fresh basil leaves or parsley",
          ],
          "instructions": Array [
            "Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato",
            "Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.",
            "Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.",
          ],
          "name": "Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref",
          "owner": 11,
          "reviews": Array [],
          "upVoteCount": 0,
          "updatedAt": "2017-11-11T09:07:34.864Z",
          "viewCount": 0,
        }
      }
    />
  </li>
  <li
    className="top-rated__item col s12 m6 l4 x4"
    key="20"
  >
    <RecipeCard
      recipe={
        Object {
          "User": Object {
            "fullname": "Email Mail",
            "id": 11,
            "username": null,
          },
          "createdAt": "2017-11-11T09:07:09.090Z",
          "description": "This is one of my 'go to' recipes ",
          "downVoteCount": 0,
          "favoriteCount": 0,
          "id": 20,
          "img_url": "http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg",
          "ingredients": Array [
            "2 tablespoons olive oil",
            "3 cloves garlic",
            "2 tablespoons capers",
            "1 tablespoon anchovy paste",
            "1 teaspoon red pepper flakes",
            "1 (24 ounce) jar tomato pasta sauce",
            "1/4 cup water",
            "1 cup good-quality crisp white wine",
            "2 (6.5 ounce) cans chopped clams, drained with juice reserved",
            "salt and pepper to taste",
            "1 pound dry spaghetti",
            "1/4 cup freshly grated Parmesan cheese",
            "Chopped fresh basil leaves or parsley",
          ],
          "instructions": Array [
            "Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato",
            "Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.",
            "Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.",
          ],
          "name": "Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref",
          "owner": 11,
          "reviews": Array [],
          "upVoteCount": 0,
          "updatedAt": "2017-11-11T09:07:09.090Z",
          "viewCount": 0,
        }
      }
    />
  </li>
  <li
    className="top-rated__item col s12 m6 l4 x4"
    key="82"
  >
    <RecipeCard
      recipe={
        Object {
          "User": Object {
            "fullname": "Email Mail",
            "id": 11,
            "username": null,
          },
          "createdAt": "2017-11-11T09:07:28.891Z",
          "description": "This is one of my 'go to' recipes ",
          "downVoteCount": 0,
          "favoriteCount": 0,
          "id": 82,
          "img_url": "http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg",
          "ingredients": Array [
            "2 tablespoons olive oil",
            "3 cloves garlic",
            "2 tablespoons capers",
            "1 tablespoon anchovy paste",
            "1 teaspoon red pepper flakes",
            "1 (24 ounce) jar tomato pasta sauce",
            "1/4 cup water",
            "1 cup good-quality crisp white wine",
            "2 (6.5 ounce) cans chopped clams, drained with juice reserved",
            "salt and pepper to taste",
            "1 pound dry spaghetti",
            "1/4 cup freshly grated Parmesan cheese",
            "Chopped fresh basil leaves or parsley",
          ],
          "instructions": Array [
            "Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato",
            "Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.",
            "Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.",
          ],
          "name": "Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref",
          "owner": 11,
          "reviews": Array [],
          "upVoteCount": 0,
          "updatedAt": "2017-11-11T09:07:28.891Z",
          "viewCount": 0,
        }
      }
    />
  </li>
  <li
    className="top-rated__item col s12 m6 l4 x4"
    key="25"
  >
    <RecipeCard
      recipe={
        Object {
          "User": Object {
            "fullname": "Email Mail",
            "id": 11,
            "username": null,
          },
          "createdAt": "2017-11-11T09:07:11.059Z",
          "description": "This is one of my 'go to' recipes ",
          "downVoteCount": 0,
          "favoriteCount": 0,
          "id": 25,
          "img_url": "http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg",
          "ingredients": Array [
            "2 tablespoons olive oil",
            "3 cloves garlic",
            "2 tablespoons capers",
            "1 tablespoon anchovy paste",
            "1 teaspoon red pepper flakes",
            "1 (24 ounce) jar tomato pasta sauce",
            "1/4 cup water",
            "1 cup good-quality crisp white wine",
            "2 (6.5 ounce) cans chopped clams, drained with juice reserved",
            "salt and pepper to taste",
            "1 pound dry spaghetti",
            "1/4 cup freshly grated Parmesan cheese",
            "Chopped fresh basil leaves or parsley",
          ],
          "instructions": Array [
            "Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato",
            "Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.",
            "Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.",
          ],
          "name": "Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref",
          "owner": 11,
          "reviews": Array [],
          "upVoteCount": 0,
          "updatedAt": "2017-11-11T09:07:11.059Z",
          "viewCount": 0,
        }
      }
    />
  </li>
</ul>
`;
